### PR TITLE
Fix(Core/Script) Shadowmeld 2

### PR DESCRIPTION
**Pull request by Hacki95**
## Changes Proposed:
-  Make it so it works like vanish, meaning you get out of combat even if the monster stands before you like here: https://www.youtube.com/watch?v=7D9-cIP6PcU

## Issues Addressed:
- Closes chromiecraft/chromiecraft#153
- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/4764

## Tests Performed:
- Fight against Yellow enemy, get out of combat
- Fight against Red enemy, only get out of combat if you have a bit of distance because you will instantly aggro again otherwise.
- Tried it with a groupmember, if used the groupmember gets aggro. After you get visible again aggro gets restored like it should.
- Tried it with a raid boss, raid boss ignors it like he should.

## How to Test the Changes:
- Make nightelf
- Fight against enemy
- Use Shadowmeld
- You will get out of combat IF the enemy is yellow OR you got a bit of distance between you and the red mob.

## Known Issues and TODO List:
- Im not sure what should happen if you are in an raid encounter and in a group. When you are alone its pretty obvious, dont end the encounter. But when you are with a group, are you supposed to drop and regain aggro or does every Boss ignore Shadowmeld? Need some info for this.

## Target Branch(es):
- [x] Master

 
## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here in the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

### CODE REVIEW

**Comment by Stifler82** [comment]
@Hacki95 
Please don't touch this spell, because in my opinion the spell is not buggy, I haven't updated the Core yet.
So I still have the version: AzerothCore rev. d20a15355c1a + 2021-03-14 19:51:16 +0000 (master branch) (Win64, Release)
Where (if I'm not mistaken) your fix isn't there.

**Video Test:** [Spell - Shadowmeld](https://youtu.be/r-jqAks-BgA)
I don't know what exactly your PR does to spell.
But from what I have been able to ascertain from my Test, **the spell works correctly!**
we avoid doing damage, if possible..

**Comment by Hacki95** [comment]
> 
> 
> @Hacki95 @FrancescoBorzi @Helias
> Please don't touch this spell, because in my opinion the spell is not buggy, I haven't updated the Core yet.
> So I still have the version: AzerothCore rev. [d20a153](https://github.com/azerothcore/azerothcore-wotlk/commit/d20a15355c1aca4942a726d2092c66918b94e9ee) + 2021-03-14 19:51:16 +0000 (master branch) (Win64, Release)
> Where (if I'm not mistaken) your fix isn't there.
> 
> **Video Test:** [Spell - Shadowmeld](https://youtu.be/r-jqAks-BgA)
> I don't know what exactly your PR does to spell.
> But from what I have been able to ascertain from my Test, **the spell works correctly!**
> we avoid doing damage, if possible..
> 
> This is part of the original report: `Night Elf's Shadowmeld ability is not working as intented. If I activate it while being attacked by a mob it doesn't drop aggro. The mob will continue to attack me instead of retreating and leaving me alone.`
> **But if you see my video, you will realize that this is not true, because the spell works perfectly!**
> 
> This spell is not an "Invisibility" Spell, it is a Stealth Spell like the Rogue Spell, although it is more like Spell Vanish.
> Spells Stealth can be detected!
> 
> **[EDIT]**
> ok Sorry, I was wrong, I'm the same version of azerothcore:master.
> So, I don't know what you need to correct now for this fix, because it seems to me that the Spell is working fine now..

I would have said the same thing and thought the spell was working as intended, but take a look this: https://www.youtube.com/watch?v=7D9-cIP6PcU which is a retail video. Try that with the current master. It won't work.

Now with this fix, this will work but only for yellow mobs, since most red mobs will still immediately attack you.

Also I'm pretty sure the old fix didn't react the right way if you were in a raid boss battle.

My main problem is that I don't have enough footage from retail to see every single edge case there is, but I am sure the way it is right now is NOT working as intended.

**Comment by Stifler82** [comment]
> I would have said the same thing and thought the spell was working as intended, but take a look this: https://www.youtube.com/watch?v=7D9-cIP6PcU which is a retail video. Try that with the current master. It won't work.
> 
> Now with this fix, this will work but only for yellow mobs, since most red mobs will still immediately attack you.
> 
> Also I'm pretty sure the old fix didn't react the right way if you were in a raid boss battle.
> 
> My main problem is that I don't have enough footage from retail to see every single edge case there is, but I am sure the way it is right now is NOT working as intended.


@Hacki95 
But you watch this video that you linked to me between 12 and 15 sec, and you will see that it is the same as my video, the player at that moment is NOT in the path of the NPCs and therefore is not detect.
[shadowmeld in 3.0/wotlk](https://youtu.be/7D9-cIP6PcU?t=12)

**Watch all my video and you will understand**
**Video Test:** [Spell - Shadowmeld](https://youtu.be/r-jqAks-BgA)

**Comment by Hacki95** [comment]
No I understand. Its just that it with the old "fix" it seems to not work in some cases where it should work. Try to recreate the video I linked in the starting zone it won't work.

**Comment by Maurowin** [comment]
I seen on your video that your char haven't right animation when u'r under shadowmeld. Normaly, u should have this animation : 
https://youtu.be/RpZlK3l_Jko?t=150

**Comment by eTzmNcbkrng** [comment]
> 
> 
> I seen on your video that your char haven't right animation when u'r under shadowmeld. Normaly, u should have this animation :
> https://youtu.be/RpZlK3l_Jko?t=150

I'm having a really hard time making out the stealth animation in your video due to the player stacking on the flight master.

Which video is incorrect?

https://www.youtube.com/watch?v=7D9-cIP6PcU
 - This is from 2008, it has no stealth animation at all, but it is from a Beta build (Tooltips have `<Alt + Left click for Feedback>` added to them)

https://youtu.be/r-jqAks-BgA
 - This is from AzerothCore, but it is a female, so different to the video you linked.

Following is a Male Night Elf Rogue in stealth from December 2009.
This is what the Male Night Elf Rogue Stealth animation should look like in 3.3.5
 - https://youtu.be/v6NAk7dpVxo?t=48

**Comment by Hacki95** [comment]
> 
> 
> > I seen on your video that your char haven't right animation when u'r under shadowmeld. Normaly, u should have this animation :
> > https://youtu.be/RpZlK3l_Jko?t=150
> 
> I'm having a really hard time making out the stealth animation in your video due to the player stacking on the flight master.
> 
> Which video is incorrect?
> 
> https://www.youtube.com/watch?v=7D9-cIP6PcU
> 
>     * This is from 2008, it has no stealth animation at all, but it is from a Beta build (Tooltips have `<Alt + Left click for Feedback>` added to them)
> 
> 
> https://youtu.be/r-jqAks-BgA
> 
>     * This is from AzerothCore, but it is a female, so different to the video you linked.
> 
> 
> Following is a Male Night Elf Rogue in stealth from December 2009.
> This is what the Male Night Elf Rogue Stealth animation should look like in 3.3.5
> 
>     * https://youtu.be/v6NAk7dpVxo?t=48

You might be right about the animation being wrong I haven't looked into that part too much tbh. The thing is there nothing in the script for the spell that tells the player to do that "emote". It could be that in the first video I linked the emote didn't trigger for whatever reason doesn't mean its never there. I would need to find additional videos of shadowmeld but the problem is there aren't many videos I found of players using shadowmeld for PVE instead of PVP.

This PR is mainly to fix the mechanic of the spell, but it seems to be rather hard to find out how the spell is actually supposed to work in certain situations because of a lack of sources.

Did you try out this PR?

**Comment by FrancescoBorzi** [comment]
what's the status of this?

**Comment by Hacki95** [comment]
> 
> 
> what's the status of this?

Still need someone to test this

**Comment by FrancescoBorzi** [comment]
@azerothcore/testers can anyone please check this?

**Comment by Sedatyf** [comment]
I'm building this PR right now. I'll keep you updated

**Comment by Sedatyf** [comment]
I tested all kind of situation as @Hacki95 described :
- A fight with a yellow enemy. Get out of combat ✔️ 

https://user-images.githubusercontent.com/59012307/114945948-92a53080-9e4a-11eb-9be1-f45440797880.mp4

- A fight with a red enemy. Get out of combat with a certain distance ✔️ 

https://user-images.githubusercontent.com/59012307/114945980-a05ab600-9e4a-11eb-80fc-09fe3d22be4b.mp4

- A fight with a red enemy and a party member in the same fight. Lose aggro and then restore it back again ✔️ 

https://user-images.githubusercontent.com/59012307/114946193-fc253f00-9e4a-11eb-927f-495b64cfb3cc.mp4

- A fight with a boss enemy in a dungeon with a party member in the same fight. Lose aggro and then restore it back again ✔️ 

https://user-images.githubusercontent.com/59012307/114946248-18c17700-9e4b-11eb-9773-bd68828882ec.mp4

- A fight with a raid boss enemy in a normal raid with a party member in the same fight. Same behavior as boss fight in dungeon ❌ (please see video attached below)
As you can see @Hacki95, Lord Marrowgar changed target and then came back to me. You said, it shouldn't work on a raid boss right?

https://user-images.githubusercontent.com/59012307/114946679-ec5a2a80-9e4b-11eb-9081-8047d1c52f23.mp4

**Comment by Knindzagxg** [comment]
@Sedatyf great videos, we can pretty much see all in them.

And actually, all of them working as supposed.
While fighting any boss, no matter is it dungeon or raid, once you are out of Shadowmeld agrro you gained is restored no matter what, which means, NPC or Boss is this situation will come back to you and attack you if you had greater % of threat on it.

Also, only 1 thing missing from testing, and if it possible, please check and make another video:
- Attack Boss and gain high threat
- Use Shadowmeld 
- Let other player Taunt the boss after you use it
- Move or cancel Shadowmeld to see what happens

In this case, due Taunt Boss should stay on another target, and threat % should be lower.

**Comment by Knindzagxg** [comment]
1 more thing that i forgot.

If you can, check out in Dungeon or Raid fights, when you are last person.
After using Shadowmeld, and there is none alive, Boss should reset without chasing you (Reseting encounter)

**Comment by Hacki95** [comment]
> 1 more thing that i forgot.
> 
> If you can, check out in Dungeon or Raid fights, when you are last person.
> After using Shadowmeld, and there is none alive, Boss should reset without chasing you (Reseting encounter)

Uhm, are you sure about that?
From my understanding raid encounters will never reset with Shadowmeld as bosses can see right through it.

**Comment by Branel** [comment]
Bosses can only see through shadowmeld if there are other alive/visible players. Lets say there is a wipe and night elf is last alive and shadowmelds, boss will not be able to find any 'visible' and resets. This has varied expansion to expansion but is the tradition way it has worked.

**Comment by Sedatyf** [comment]
I can test all of that but which behavior are we looking then? Because, you guys do not agree with each other.

**Comment by Hacki95** [comment]
> 
> 
> Bosses can only see through shadowmeld if there are other alive/visible players. Lets say there is a wipe and night elf is last alive and shadowmelds, boss will not be able to find any 'visible' and resets. This has varied expansion to expansion but is the tradition way it has worked.

Do you have any reference for that? My main problem is finding sources which back this up.



> 
> 
> I can test all of that but which behavior are we looking then? Because, you guys do not agree with each other.

No he is right, the boss part for raids is the only situation I am not 100% sure about how this spell should work. It might need some changes here but I need a source.

**Comment by Branel** [comment]
https://flyv.typepad.com/warcraft/2009/01/the-many-uses-of-shadowmeld.html

**Comment by Sedatyf** [comment]
> No he is right, the boss part for raids is the only situation I am not 100% sure about how this spell should work. It might need some changes here but I need a source.

Alright. Feel free to tag me here or in Discord when you think it's ready and I'll test it again

**Comment by Hacki95** [comment]
> 
> 
> https://flyv.typepad.com/warcraft/2009/01/the-many-uses-of-shadowmeld.html

"Wipe recovery. If the raid is about to wipe it occasionally works to Shadowmeld in a quiet corner, let the boss reset, then rez people."

This is the only point referencing this, the problem I see is that if you are alone against he boss he is probably right inside your face, so reset isn't possible in those occasions. I will try to get away from a boss and see if it works under certain circumstances, but it looks like it shouldn't work always.

One thing I am also not 100% sure about is: "Handing aggro to another tank. I'm using Shadowmeld a fair amount when tanking. It's useful on fights where you have to hand-off aggro to another tank, like on Gluth. Sure the tank can just taunt off you, but if the taunt misses or they're not paying attention you can just hand the other tank aggro. I also shadowmeld trash on to my fellow tanks when I'm done bear AoEing and want to go kitty. You want to be very sure that another tank is #2 on aggro before you do it and don't shadowmeld a lot of damage onto another tank if your healers aren't ready to switch."

On one hand the author suggests it is possible to completly reset your own threat by using Shadowmeld, on the other hand the tooltip of the spell reads that when Shadowmeld is canceled that you should regain your threat. This could mean that:
1) The tooltip is wrong
2) The author meant that you had to stay in Shadowmeld for a while for the other tank to "outthreat" you, then you can cancel it without getting aggro back.

Can anyone confirm/deny this?

**Comment by Branel** [comment]
Tooltips are almost always missleading and there are many many cases where tooltips are completly wrong and you have to look for videos, forums posts and external sources to find how a spell actuall worked - have dealt with that before.

As much as I understand is that when shadowmeld is **active** then you have 0 threat, and the moment you have cancelled shadowmeld you will get you threat back. 

Incase im not clear enogh
1. You have 250% threat -> cast shadowmeld
2. While shadowmeld is active you have 0% threat
3. When shadowmeld is cancelled you get put back on the threat meter (by this time the other tank already got a fair chance to get ahold of the aggro/overcome your amount of threat)

If you still dont believe shadowmeld can be used to reset wipes I am happy to find more valuable sources, just let me know.

To clarify, when the boss gets reset and he is pathing way to his origin point, you cannot pull him so in most cases even if you are right infront of you as long as he is just not about to swing at you or as long as you're not in AoE or generally taking damage as soon as you shadowmeld you will get put out of combat and he will path to his origin point and while he is moving back you are out of combat and you get to run away from him.

What he actually meant with it 'not working always' is that some boss encounters have constant ongoing damage so that you cannot reset the boss if you are constantly taking damage.

Previously I was on my phone and couldn't give enough detailed information, hope this helps.

**Comment by Sedatyf** [comment]
Okay so I'll need to do one more test to check if I can reset a boss with Shadowmeld and I think we have tested every behavior

**Comment by Hacki95** [comment]
> 
> 
> Okay so I'll need to do one more test to check if I can reset a boss with Shadowmeld and I think we have tested every behavior

Yes please try that but I am pretty sure I will need to change something in the code for this to work.
It might work right now if you have some distance to the boss, if you are right in his face it will probably not work, but this should be right, isn't it @Branel ? (I tried with Marrowgar before and I wasn't able to reset the raid when he was right in my face).

**Comment by Branel** [comment]
Sounds about right, marrowgar is a fight you cant reset outside of a regular wipe afaik because of how the boss functions.

**Comment by Branel** [comment]
I have tested and currently this branch does not reset the boss if you use shadowmeld and you are the only party member alive/visible. Distance from a boss should not matter but for good measure I let the boss be a considerable distance away from my character and he walked right up to me and swinged at my character,[ see gif as example.](https://gfycat.com/velvetyinformalcopepod)




I have not yet tested the other features but looking at the previous videos everything seems to be in order, even the fact that it resets your aggro on shadowmeld. If inquired I can do further testing.

Overall good job though.[](url)

**Comment by Sedatyf** [comment]
Ah @Branel you were quicker than me. I just did a test too and I can confirm that Shadowmeld doesn't reset boss fight. Please see video attached below

https://user-images.githubusercontent.com/59012307/115054517-94213800-9ee0-11eb-9c71-353689428198.mp4

**Comment by Branel** [comment]
> Ah @Branel you were quicker than me. I just did a test too and I can confirm that Shadowmeld doesn't reset boss fight. Please see video attached below
> 
>  2021-04-16.18-18-17.mp4

its fine good job

**Comment by Hacki95** [comment]
Updated it, if one of you could try it again that would be great!

**Comment by Sedatyf** [comment]
> Updated it, if one of you could try it again that would be great!

I'm looking at it. I'll keep you updated

**Comment by Sedatyf** [comment]
We can leave combat when a raid boss fight if the night elf is the last one alive ✔️ 

https://user-images.githubusercontent.com/59012307/115085265-7ff23080-9f0a-11eb-9f25-7e3ae0c581a0.mp4

I think we're done here!

**Comment by Hacki95** [comment]
@Knindzagxg Can you take another look now?

**Comment by Knindzagxg** [comment]
Yeah, this looks good man.

Just one more thing to test:
- Keep attacking boss till you gain higher threat
- Use Shadowmeld and let other tank taunt / gain threat
- Leave Shadowmeld to see what's happening (You will restore some threat, but Boss should stay on other tank)

If that working, we can merge xP

**Comment by Sedatyf** [comment]
I'll test it today! I'll comment when done

**Comment by Knindzagxg** [comment]
Cheers mate.
Video would be great, again xD

File: src/server/scripts/Spells/spell_generic.cpp [hunk 1]
```diff
@@ -5111,13 +5111,10 @@ class spell_gen_shadowmeld : public SpellScriptLoader
...
             }
 
-            if (!caster->GetInstanceScript() || !caster->GetInstanceScript()->IsEncounterInProgress()) //Don't leave combat if you are in combat with a boss
-            {
-                if (!instant_exit) {
-                    caster->getHostileRefManager().deleteReferences(); // exit combat after 6 seconds
-                }
-                else caster->CombatStop(); // isn't necessary to call AttackStop because is just called in CombatStop
+            if (instant_exit) {
+                caster->getHostileRefManager().deleteReferences();
```
Comment (FrancescoBorzi) [inline src/server/scripts/Spells/spell_generic.cpp:5115]: the bracket should be in a new line

**Comment by Sedatyf** [comment]
It seems I can't build this PR in my Visual Studio. It says that ``Banner.cpp`` is missing. Does this need to be merged with master before testing?

**Comment by FrancescoBorzi** [comment]
> It seems I can't build this PR in my Visual Studio. It says that `Banner.cpp` is missing. Does this need to be merged with master before testing?

@Sedatyf that's weird because the build is passing here

image

including the windows one. Maybe you have to do a clean build?

**Comment by Sedatyf** [comment]
Unfortunately, I couldn't test it today. I'll try tomorrow and I'll keep you updated as usual

**Comment by Sedatyf** [comment]
> Yeah, this looks good man.
> 
> Just one more thing to test:
> 
> * Keep attacking boss till you gain higher threat
> * Use Shadowmeld and let other tank taunt / gain threat
> * Leave Shadowmeld to see what's happening (You will restore some threat, but Boss should stay on other tank)
> 
> If that working, we can merge xP

Okay so we made two tests : one like you described, and the other one, tank just taunt but doesn't get higher threat, so boss will come back on night elf. And everything work fine!

https://user-images.githubusercontent.com/59012307/115305675-f9865a80-a166-11eb-8cf9-caec5e33c6a4.mp4


https://user-images.githubusercontent.com/59012307/115305686-fd19e180-a166-11eb-9f99-f8f61f40a470.mp4

Congrats to @Hacki95!

We just need a review here now 😄

**Comment by Hacki95** [comment]
Thanks a lot @Sedatyf, excellent job! I will fix the bracket thing and ping an admin tomorrow.

**Comment by Helias** [comment]
very nice :rocket:

**Review by Helias — APPROVED** [review]

**Comment by Knindzagxg** [comment]
Perfect xD
Cheers for videos, and great job Hacki xD

**Review by Knindzagxg — APPROVED** [review]

**Pull request merged**